substrate/client/network/sync/src/chain_sync.rs

```diff
@@ -0,0 +1,3406 @@
+	Block: BlockT,
+	T: HeaderMetadata<Block, Error = sp_blockchain::Error> + ?Sized,
+{
+	if base == block {
+		return Ok(false)
+	}
+
+	let ancestor = sp_blockchain::lowest_common_ancestor(client, *block, *base)?;
+
+	Ok(ancestor.hash == *base)
+}
+
+/// Validate that the given `blocks` are correct.
+/// Returns the number of the first block in the sequence.
+///
+/// It is expected that `blocks` are in ascending order.
+fn validate_blocks<Block: BlockT>(
+	blocks: &Vec<BlockData<Block>>,
+	who: &PeerId,
+	request: Option<BlockRequest<Block>>,
+) -> Result<Option<NumberFor<Block>>, BadPeer> {
+	if let Some(request) = request {
+		if Some(blocks.len() as _) > request.max {
+			debug!(
+				target: LOG_TARGET,
+				"Received more blocks than requested from {}. Expected in maximum {:?}, got {}.",
+				who,
+				request.max,
+				blocks.len(),
+			);
+
+			return Err(BadPeer(*who, rep::NOT_REQUESTED))
+		}
+
+		let block_header =
+			if request.direction == Direction::Descending { blocks.last() } else { blocks.first() }
+				.and_then(|b| b.header.as_ref());
+
+		let expected_block = block_header.as_ref().map_or(false, |h| match request.from {
+			FromBlock::Hash(hash) => h.hash() == hash,
+			FromBlock::Number(n) => h.number() == &n,
+		});
+
+		if !expected_block {
+			debug!(
+				target: LOG_TARGET,
+				"Received block that was not requested. Requested {:?}, got {:?}.",
+				request.from,
+				block_header,
+			);
+
+			return Err(BadPeer(*who, rep::NOT_REQUESTED))
+		}
+
+		if request.fields.contains(BlockAttributes::HEADER) &&
+			blocks.iter().any(|b| b.header.is_none())
+		{
+			trace!(
+				target: LOG_TARGET,
+				"Missing requested header for a block in response from {who}.",
+			);
+
+			return Err(BadPeer(*who, rep::BAD_RESPONSE))
+		}
+
+		if request.fields.contains(BlockAttributes::BODY) && blocks.iter().any(|b| b.body.is_none())
+		{
+			trace!(
+				target: LOG_TARGET,
+				"Missing requested body for a block in response from {who}.",
+			);
+
+			return Err(BadPeer(*who, rep::BAD_RESPONSE))
+		}
+	}
+
+	for b in blocks {
+		if let Some(header) = &b.header {
+			let hash = header.hash();
+			if hash != b.hash {
+				debug!(
+					target:LOG_TARGET,
+					"Bad header received from {}. Expected hash {:?}, got {:?}",
+					who,
+					b.hash,
+					hash,
+				);
+				return Err(BadPeer(*who, rep::BAD_BLOCK))
+			}
+		}
+		if let (Some(header), Some(body)) = (&b.header, &b.body) {
+			let expected = *header.extrinsics_root();
+			let got = HashingFor::<Block>::ordered_trie_root(
+				body.iter().map(Encode::encode).collect(),
+				sp_runtime::StateVersion::V0,
+			);
+			if expected != got {
+				debug!(
+					target:LOG_TARGET,
+					"Bad extrinsic root for a block {} received from {}. Expected {:?}, got {:?}",
+					b.hash,
+					who,
+					expected,
+					got,
+				);
+				return Err(BadPeer(*who, rep::BAD_BLOCK))
+			}
+		}
+	}
+
+	Ok(blocks.first().and_then(|b| b.header.as_ref()).map(|h| *h.number()))
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
If you are at it, why not also move them into their own file?

Reply:
Done